### PR TITLE
Map serialization

### LINKED ISSUE #1: How to store/retrieve puzzles?
Hi, 

I've been working on a simple sudoku app from a couple weeks now and I thank you for this api! it's been really useful and it seems to work flawless!

one think that I haven't been able to figure out is how would you reopen an already started puzzled (The user left the app and wants to continue later)? is that something that can be done with the current state of your api?

Thanks in advance, overall this is great!

### PR DESCRIPTION
**Pull request by JoseGeorges8**
- Added serialization for puzzle and its dependent models PuzzleOptions, Grid, Cell and Position.
- Removed the cell interface from the Cell class since it wouldn't let me create a private constructor, probably not the best choice but worked for me (Hoping for a suggestion on how to approach this differently). 
- Added a new property in Puzzle to keep track of time when making it a map

### CODE REVIEW

**Comment by JoseGeorges8** [comment]
@AlvinRamoutar I've updated the PR

**Comment by AlvinRamoutar** [comment]
Great; will review and get back to you soon

File: lib/src/Puzzle.dart [hunk 1]
```diff
@@ -22,6 +22,7 @@ class Puzzle {
   Patterner _patterner;
   PuzzleOptions _options;
   Stopwatch _stopwatch;
+  int _timeElapsedInSeconds = 0; // holds the elapsed time for when getting converted to a map
```
Comment (AlvinRamoutar) [inline lib/src/Puzzle.dart:27]: Trying to figure out where/why this exists.
I'm only seeing one write to it, in [Line 47](https://github.com/JoseGeorges8/sudoku-api/blob/95b4dff6533dbff802540617de1e1386a2b328e3/lib/src/Puzzle.dart#L47) of `Puzzle`.

File: lib/src/Puzzle.dart [hunk 2]
```diff
@@ -34,8 +35,51 @@ class Puzzle {
...
+    "options": options() == null ? null : options().toMap(),
+    "time_elapsed_in_seconds": getTimeElapsed().inSeconds
+  };
+
+
   /// Generates a new puzzle using parameters set in [_options]
   Future<bool> generate() async {
+
+    if(_board != null)
+      throw('Board already generated');
```
Comment (AlvinRamoutar) [inline lib/src/Puzzle.dart]: Good catch; could we see an exception implementation in `/src/logic/SudokuException`?
Something simple; just planning for the future.
I.e.:

```
/// Throw whenever [Puzzle] experiences a generation related issue
class GenerationException implements Exception {
  String cause;
  GenerationException(this.cause);
}
```

File: lib/src/Puzzle.dart [hunk 3]
```diff
@@ -103,8 +147,11 @@ class Puzzle {
   /// I can only make these comments so interesting and no more :l
   void startStopwatch() => _stopwatch.start();
   void stopStopwatch() => _stopwatch.stop();
-  Duration getTimeElapsed() => _stopwatch.elapsed;
+  /// Add the time elapsed in case the game is being reloaded from map/storage
+  Duration getTimeElapsed() => Duration(seconds: _timeElapsedInSeconds) + _stopwatch.elapsed;
```
Comment (AlvinRamoutar) [inline lib/src/Puzzle.dart:153]: If [Line 47](https://github.com/JoseGeorges8/sudoku-api/blob/95b4dff6533dbff802540617de1e1386a2b328e3/lib/src/Puzzle.dart#L47) ends up being redundant, we can replace with:

```suggestion
   Duration getTimeElapsed() => _stopwatch.elapsed;
```

File: lib/src/models/Cell.dart [hunk 4]
```diff
@@ -5,7 +5,13 @@ import 'ICell.dart';
 import 'Position.dart';
 
 /// Represents a single cell in a 9x9 grid
-class Cell extends ICell {
+class Cell{
```
Comment (AlvinRamoutar) [inline lib/src/models/Cell.dart]:
```suggestion
class Cell extends ICell {
```

_Part of the 'let's keep ICell' movement ;)_

File: lib/src/models/Cell.dart [hunk 5]
```diff
@@ -5,7 +5,13 @@ import 'ICell.dart';
...
+  /// Position of this cell in a 9x9 grid
+  final Position position;
+
+  /// Whether or not this cell's value has been changed since grid generation
+  bool isPristine;
+
```
Comment (AlvinRamoutar) [inline lib/src/models/Cell.dart]:
```suggestion

```

_Part of the 'let's keep ICell' movement ;)_

File: lib/src/models/Cell.dart [hunk 6]
```diff
@@ -23,13 +29,42 @@ class Cell extends ICell {
 
   /// Constructs new Cell at [position] with optional [_value]
   /// If value is provided, then cell is flagged as prefilled
-  Cell(position, [this._value = 0]) : super(position) {
+  Cell(this.position, [this._value = 0]){
```
Comment (AlvinRamoutar) [inline lib/src/models/Cell.dart]:
```suggestion
  Cell(position, [this._value = 0]) : super(position, true){
```

_Part of the 'let's keep ICell' movement ;)_

File: lib/src/models/Cell.dart [hunk 7]
```diff
@@ -23,13 +29,42 @@ class Cell extends ICell {
...
   }
 
+  Cell._({int value,
+    bool isPrefill,
+    bool isValid,
+    bool isMarkup,
+    bool isPristine,
+    this.position}) : this._isPrefill = isPrefill, this._value = value, this._isValid = isValid, this._isMarkup = isMarkup, this.isPristine = isPristine {
+    _onChange = new StreamController.broadcast();
+  }
```
Comment (AlvinRamoutar) [inline lib/src/models/Cell.dart]:
```suggestion
  Cell._({int value,
    bool isPrefill,
    bool isValid,
    bool isMarkup,
    bool isPristine,
    Position position}) : this._isPrefill = isPrefill, this._value = value, this._isValid = isValid, this._isMarkup = isMarkup, super(position, isPristine) {
    _onChange = new StreamController.broadcast();
  }
```

_Part of the 'let's keep ICell' movement ;)_

**Review by AlvinRamoutar — CHANGES_REQUESTED** [review]
Hey @JoseGeorges8 ,

I've made some suggestions, a major one being how we can handle ICell.
You can easily find those ICell-related suggestions by their footer: 
_Part of the 'let's keep ICell' movement ;)_

TO implement these ICell changes, we'd have to change ICell's [constructor](https://github.com/AlvinRamoutar/sudoku-api/blob/111da91d4bda90b4f82d824eb33a254f5aaa222d/lib/src/models/ICell.dart#L12) to the following:
`ICell(this.position, this.isPristine);`

I've ran `serialization_test` successfully.

Let me know what you think - thank you for the mapping contributions!

File: lib/src/Puzzle.dart (hunk 1 above)
Comment (JoseGeorges8) [reply lib/src/Puzzle.dart:27]: So to get the time elapsed you're using a Stopwatch, and looking through the class I couldn't find a way to add/set time to the stopwatch itself, and if someone wants to continue the puzzle later, the time that already passed should be recorded therefore I thought this variable could hold the timeElapsed up to the point where the map was created/ready to be stored.


- [Line 73](https://github.com/JoseGeorges8/sudoku-api/blob/95b4dff6533dbff802540617de1e1386a2b328e3/lib/src/Puzzle.dart#L73) Assigns the value to the map `time_elapsed_in_seconds`
- [Line 151](https://github.com/JoseGeorges8/sudoku-api/blob/95b4dff6533dbff802540617de1e1386a2b328e3/lib/src/Puzzle.dart#L151) Now gives you the current time passed + the timeElapsed previously
- [Line 66](https://github.com/JoseGeorges8/sudoku-api/blob/95b4dff6533dbff802540617de1e1386a2b328e3/lib/src/Puzzle.dart#L66) retrieves the value from the `time_elapsed_in_seconds` map property and assigns it to the puzzle `_timeElapsedInSeconds ` property

Hopefully this clarifies the where/why property!

File: lib/src/Puzzle.dart (hunk 2 above)
Comment (JoseGeorges8) [reply lib/src/Puzzle.dart]: Yeah I'll create and update the code to include this

File: lib/src/Puzzle.dart (hunk 3 above)
Comment (JoseGeorges8) [reply lib/src/Puzzle.dart:153]: Check the comment above about the [Line 47](https://github.com/JoseGeorges8/sudoku-api/blob/95b4dff6533dbff802540617de1e1386a2b328e3/lib/src/Puzzle.dart#L47)

**Comment by JoseGeorges8** [comment]
@AlvinRamoutar  Changes have been reviewed and applied 👍Thanks for the suggestions!

**Review by AlvinRamoutar — APPROVED** [review]
Sweet!
Thanks for the clarification regarding `_timeElapsedInSeconds `, let's leave it like that for now until there's a better way we can handle it in the future.
Everything else looks good, `generation_test` and `serialization_test` were successful.

Proceeding with merging - thank you for your contribution @JoseGeorges8 ! 😄

**Pull request merged**